File: backend/src/services/mcp/apiKeyManager.ts
/**
 * ApiKeyManager - MCP Server API Key CRUD + 加密存储
 *
 * 管理 MCP Server 的 API Key 生命周期：创建、验证、撤销、列表
 * 使用 CryptoService (AES-256) 加密存储，async-mutex 防止并发竞态
 *
 * Requirements: 10.1, 10.2, 10.3, 10.4
 */

import { Mutex } from 'async-mutex';
import { v4 as uuidv4 } from 'uuid';
import * as crypto from 'crypto';
import { cryptoService } from '../ai/cryptoService';
import { logger } from '../../utils/logger';
import type { DataStore } from '../dataStore';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface McpApiKey {
  id: string;
  keyHash: string;           // CryptoService 加密后的 Key
  keyPrefix: string;         // 明文前 8 位（如 mcp_a1b2），用于验证时快速过滤
  tenantId: string;
  role: 'admin' | 'operator' | 'viewer';
  label: string;
  status: 'active' | 'revoked';
  createdAt: number;
  revokedAt?: number;
}

export interface SecurityContext {
  tenantId: string;
  role: 'admin' | 'operator' | 'viewer';
  apiKeyId: string;
  clientId?: string;
}

// ─── ApiKeyManager ───────────────────────────────────────────────────────────

export class ApiKeyManager {
  private dataStore: DataStore | null = null;
  private writeMutex: Mutex;

  constructor() {
    this.writeMutex = new Mutex();
  }

  /**
   * 注入 DataStore（PostgreSQL），启用持久化
   */
  setDataStore(dataStore: DataStore): void {
    this.dataStore = dataStore;
    logger.info('[ApiKeyManager] DataStore injected, using PostgreSQL for API key storage');
  }

  /**
   * 读取所有 Key（从 PostgreSQL）
   */
  private async readKeys(): Promise<McpApiKey[]> {
    if (!this.dataStore) {
      logger.warn('[ApiKeyManager] No DataStore configured, returning empty key list');
      return [];
    }
    const rows = await this.dataStore.query<{
      id: string;
      key_hash: string;
      key_prefix: string;
      tenant_id: string;
      role: string;
      label: string;
      status: string;
      created_at: number;
      revoked_at: number | null;
    }>('SELECT id, key_hash, key_prefix, tenant_id, role, label, status, created_at, revoked_at FROM mcp_api_keys');
    return rows.map(r => ({
      id: r.id,
      keyHash: r.key_hash,
      keyPrefix: r.key_prefix,
      tenantId: r.tenant_id,
      role: r.role as McpApiKey['role'],
      label: r.label,
      status: r.status as McpApiKey['status'],
      createdAt: r.created_at,
      revokedAt: r.revoked_at ?? undefined,
    }));
  }

  /**
   * 创建新的 API Key
   * 返回明文 Key（仅此一次）和元数据
   */
  async createKey(
    tenantId: string,
    role: McpApiKey['role'],
    label: string
  ): Promise<{ key: string; metadata: Omit<McpApiKey, 'keyHash'> }> {
    if (!this.dataStore) {
      throw new Error('ApiKeyManager: DataStore not configured');
    }

    return this.writeMutex.runExclusive(async () => {
      // 生成明文 Key: mcp_ 前缀 + UUID
      const rawKey = `mcp_${uuidv4().replace(/-/g, '')}`;
      const keyPrefix = rawKey.substring(0, 8);
      const keyHash = cryptoService.encrypt(rawKey);
      const id = uuidv4();
      const now = Date.now();

      await this.dataStore!.execute(
        `INSERT INTO mcp_api_keys (id, key_hash, key_prefix, tenant_id, role, label, status, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        [id, keyHash, keyPrefix, tenantId, role, label, 'active', now]
      );

      logger.info(`[ApiKeyManager] Created API Key: id=${id}, tenant=${tenantId}, role=${role}, label=${label}`);

      const metadata: Omit<McpApiKey, 'keyHash'> = {
        id, keyPrefix, tenantId, role, label, status: 'active', createdAt: now,
      };
      return { key: rawKey, metadata };
    });
  }

  /**
   * 撤销 API Key — 立即生效
   */
  async revokeKey(keyId: string): Promise<void> {
    if (!this.dataStore) {
      throw new Error('ApiKeyManager: DataStore not configured');
    }

    return this.writeMutex.runExclusive(async () => {
      const key = await this.dataStore!.queryOne<{ id: string; status: string }>(
        'SELECT id, status FROM mcp_api_keys WHERE id = $1',
        [keyId]
      );

      if (!key) {
        throw new Error(`API Key not found: ${keyId}`);
      }

      if (key.status === 'revoked') {
        logger.warn(`[ApiKeyManager] Key already revoked: ${keyId}`);
        return;
      }

      await this.dataStore!.execute(
        'UPDATE mcp_api_keys SET status = $1, revoked_at = $2 WHERE id = $3',
        ['revoked', Date.now(), keyId]
      );

      logger.info(`[ApiKeyManager] Revoked API Key: id=${keyId}`);
    });
  }

  /**
   * 列出所有 Key 元数据（排除 keyHash）
   */
  async listKeys(): Promise<Omit<McpApiKey, 'keyHash'>[]> {
    const keys = await this.readKeys();
    return keys.map(({ keyHash: _, ...rest }) => rest);
  }

  /**
   * 验证 API Key — 恒定时间比较，防定时攻击
   * 返回 SecurityContext 或 null
   */
  async validateKey(rawKey: string): Promise<SecurityContext | null> {
    if (!rawKey || typeof rawKey !== 'string') {
      return null;
    }

    const keys = await this.readKeys();
    const prefix = rawKey.substring(0, 8);

    // 按 keyPrefix 快速过滤 active 候选集
    const candidates = keys.filter(
      k => k.status === 'active' && k.keyPrefix === prefix
    );

    for (const candidate of candidates) {
      try {
        const decrypted = cryptoService.decrypt(candidate.keyHash);
        // 恒定时间比较
        const a = Buffer.from(rawKey, 'utf-8');
        const b = Buffer.from(decrypted, 'utf-8');
        if (a.length === b.length && crypto.timingSafeEqual(a, b)) {
          return {
            tenantId: candidate.tenantId,
            role: candidate.role,
            apiKeyId: candidate.id,
          };
        }
      } catch {
        // 解密失败，跳过此候选
        continue;
      }
    }

    return null;
  }
}

// ─── 单例 ────────────────────────────────────────────────────────────────────

export const apiKeyManager = new ApiKeyManager();

File: opsevo-python/src/opsevo/services/mcp/api_key_manager.py
"""
ApiKeyManager — MCP Server API Key CRUD + 加密存储

管理 MCP Server 的 API Key 生命周期：创建、验证、撤销、列表。
使用 CryptoService (AES-256) 加密存储，asyncio.Lock 防止并发竞态。
"""

from __future__ import annotations

import asyncio
import hmac
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

import structlog

logger = structlog.get_logger(__name__)


@dataclass
class McpApiKey:
    """MCP API Key 元数据。"""

    id: str
    key_hash: str
    key_prefix: str
    tenant_id: str
    role: str  # admin | operator | viewer
    label: str
    status: str  # active | revoked
    created_at: str  # ISO 8601 timestamp
    revoked_at: str | None = None


@dataclass
class SecurityContext:
    """安全上下文。"""

    tenant_id: str
    role: str
    api_key_id: str
    client_id: str | None = None


class ApiKeyManager:
    """MCP API Key 管理器。"""

    def __init__(self, crypto_service: Any = None) -> None:
        self._crypto = crypto_service
        self._datastore: Any = None
        self._lock = asyncio.Lock()

    def set_datastore(self, datastore: Any) -> None:
        self._datastore = datastore
        logger.info("ApiKeyManager: DataStore injected")

    # ------------------------------------------------------------------
    # 读取
    # ------------------------------------------------------------------

    async def _read_keys(self) -> list[McpApiKey]:
        if not self._datastore:
            return []
        rows = await self._datastore.query(
            "SELECT id, key_hash, key_prefix, tenant_id, role, label, status, created_at, revoked_at "
            "FROM mcp_api_keys"
        )
        return [
            McpApiKey(
                id=r["id"],
                key_hash=r["key_hash"],
                key_prefix=r["key_prefix"],
                tenant_id=r["tenant_id"],
                role=r["role"],
                label=r["label"],
                status=r["status"],
                created_at=r["created_at"],
                revoked_at=r.get("revoked_at"),
            )
            for r in rows
        ]

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    async def create_key(
        self, tenant_id: str, role: str, label: str
    ) -> dict[str, Any]:
        """创建新 API Key，返回明文 key（仅此一次）和元数据。"""
        if not self._datastore:
            raise RuntimeError("ApiKeyManager: DataStore not configured")

        async with self._lock:
            raw_key = f"mcp_{uuid.uuid4().hex}"
            key_prefix = raw_key[:8]
            key_hash = self._encrypt(raw_key)
            key_id = str(uuid.uuid4())
            now = datetime.now(timezone.utc)

            await self._datastore.execute(
                "INSERT INTO mcp_api_keys (id, key_hash, key_prefix, tenant_id, role, label, status, created_at) "
                "VALUES ($1, $2, $3, $4, $5, $6, $7, $8)",
                [key_id, key_hash, key_prefix, tenant_id, role, label, "active", now],
            )
            logger.info("API Key created", id=key_id, tenant=tenant_id, role=role)
            return {
                "key": raw_key,
                "metadata": {
                    "id": key_id,
                    "keyPrefix": key_prefix,
                    "tenantId": tenant_id,
                    "role": role,
                    "label": label,
                    "status": "active",
                    "createdAt": now.isoformat(),
                },
            }

    async def revoke_key(self, key_id: str) -> None:
        if not self._datastore:
            raise RuntimeError("ApiKeyManager: DataStore not configured")
        async with self._lock:
            row = await self._datastore.query_one(
                "SELECT id, status FROM mcp_api_keys WHERE id = $1", [key_id]
            )
            if not row:
                raise ValueError(f"API Key not found: {key_id}")
            if row["status"] == "revoked":
                return
            await self._datastore.execute(
                "UPDATE mcp_api_keys SET status = $1, revoked_at = $2 WHERE id = $3",
                ["revoked", datetime.now(timezone.utc), key_id],
            )
            logger.info("API Key revoked", id=key_id)

    async def list_keys(self) -> list[dict[str, Any]]:
        keys = await self._read_keys()
        return [
            {
                "id": k.id,
                "keyPrefix": k.key_prefix,
                "tenantId": k.tenant_id,
                "role": k.role,
                "label": k.label,
                "status": k.status,
                "createdAt": k.created_at,
                "revokedAt": k.revoked_at,
            }
            for k in keys
        ]

    async def validate_key(self, raw_key: str) -> SecurityContext | None:
        """验证 API Key，返回 SecurityContext 或 None。"""
        if not raw_key or not self._datastore:
            return None
        prefix = raw_key[:8]
        rows = await self._datastore.query(
            "SELECT id, key_hash, tenant_id, role FROM mcp_api_keys "
            "WHERE key_prefix = $1 AND status = 'active'",
            [prefix],
        )
        for r in rows:
            try:
                decrypted = self._decrypt(r["key_hash"])
                if hmac.compare_digest(raw_key, decrypted):
                    return SecurityContext(
                        tenant_id=r["tenant_id"],
                        role=r["role"],
                        api_key_id=r["id"],
                    )
            except Exception:
                logger.warning("validate_key_decrypt_failed", key_id=r["id"], exc_info=True)
                continue
        return None

    # ------------------------------------------------------------------
    # 加密辅助
    # ------------------------------------------------------------------

    def _encrypt(self, value: str) -> str:
        if self._crypto:
            return self._crypto.encrypt(value)
        raise RuntimeError(
            "CryptoService is required to create API keys. "
            "SHA256 fallback removed because one-way hash makes keys unvalidatable."
        )

    def _decrypt(self, value: str) -> str:
        if self._crypto:
            return self._crypto.decrypt(value)
        raise RuntimeError("Cannot decrypt without CryptoService")
